client/my-sites/checkout/checkout/web-payment-box.jsx

```diff
@@ -0,0 +1,517 @@
+			return {
+				default: false,
+				disabled: true,
+				text: translate( 'Completing your purchase', {
+					context: 'Loading state on /checkout',
+				} ),
+			};
+		};
+
+		switch ( transactionStep.name ) {
+			case BEFORE_SUBMIT:
+				return defaultState();
+
+			case INPUT_VALIDATION:
+				if ( transactionStep.error ) {
+					return defaultState();
+				}
+
+				return ongoingState();
+
+			case SUBMITTING_PAYMENT_KEY_REQUEST:
+				return ongoingState();
+
+			case RECEIVED_PAYMENT_KEY_RESPONSE:
+				if ( transactionStep.error ) {
+					return defaultState();
+				}
+
+				return ongoingState();
+
+			case SUBMITTING_WPCOM_REQUEST:
+				return completingState();
+
+			case RECEIVED_WPCOM_RESPONSE:
+				if ( transactionStep.error || ! transactionStep.data.success ) {
+					return defaultState();
+				}
+
+				return completingState();
+
+			default:
+				throw new Error( 'Unknown transaction step: ' + transactionStep.name );
+		}
+	};
+
+	/**
+	 * @return {PaymentRequest} A configured payment request object.
+	 */
+	getPaymentRequestForApplePay = () => {
+		const { cart, translate } = this.props;
+
+		const supportedPaymentMethods = [
+			{
+				supportedMethods: WEB_PAYMENT_APPLE_PAY_METHOD,
+				data: {
+					version: 3,
+					merchantIdentifier: APPLE_PAY_MERCHANT_IDENTIFIER,
+					merchantCapabilities: [ 'supports3DS', 'supportsCredit', 'supportsDebit' ],
+					supportedNetworks: SUPPORTED_NETWORKS,
+					countryCode: 'US',
+				},
+			},
+		];
+		const paymentDetails = {
+			total: {
+				label: translate( 'Total' ),
+				amount: {
+					currency: cart.currency,
+					value: cart.total_cost + '',
+				},
+			},
+			displayItems: cart.products.map( product => {
+				return {
+					label: product.product_name,
+					amount: {
+						currency: product.currency,
+						value: product.cost + '',
+					},
+				};
+			} ),
+		};
+
+		const paymentRequest = new PaymentRequest(
+			supportedPaymentMethods,
+			paymentDetails,
+			PAYMENT_REQUEST_OPTIONS
+		);
+
+		const environment = 'production' === config( 'env' ) ? undefined : 'sandbox';
+
+		paymentRequest.onmerchantvalidation = event => {
+			wpcom
+				.undocumented()
+				.applePayMerchantValidation( event.validationURL, environment )
+				.then( json => event.complete( json ) )
+				.catch( error => {
+					debug( 'onmerchantvalidation error' );
+					debug( error );
+				} );
+		};
+
+		return paymentRequest;
+	};
+
+	/**
+	 * @return {PaymentRequest} A configured payment request object.
+	 */
+	getPaymentRequestForBasicCard = () => {
+		const { cart, translate } = this.props;
+
+		const supportedPaymentMethods = [
+			{
+				supportedMethods: WEB_PAYMENT_BASIC_CARD_METHOD,
+				data: {
+					supportedNetworks: SUPPORTED_NETWORKS,
+				},
+			},
+		];
+		const paymentDetails = {
+			total: {
+				label: translate( 'Total' ),
+				amount: {
+					currency: cart.currency,
+					value: cart.total_cost,
+				},
+			},
+			displayItems: cart.products.map( product => {
```

Comment:
Should we make sure `cart.products` exists and is an array?

Reply:
I like to live dangerously!